crates/spirv-std/src/textures.rs

```diff
@@ -52,58 +150,41 @@ impl Image2d {
 }
 
 #[allow(unused_attributes)]
-#[spirv(image(
-    // sampled_type is hardcoded to f32 for now
-    dim = "Dim2D",
-    depth = 0,
-    arrayed = 1,
-    multisampled = 0,
-    sampled = 1,
-    image_format = "Unknown"
-))]
+#[spirv(sampler)]
 #[derive(Copy, Clone)]
-pub struct Image2dArray {
+pub struct Sampler {
     _x: u32,
 }
 
-impl Image2dArray {
-    pub fn sample(&self, sampler: Sampler, coord: Vec3A) -> Vec4 {
-        #[cfg(not(target_arch = "spirv"))]
-        {
-            let _ = sampler;
-            let _ = coord;
-            panic!("Image sampling not supported on CPU");
-        }
-        #[cfg(target_arch = "spirv")]
-        unsafe {
-            let mut result = Default::default();
-            asm!(
-                "%typeSampledImage = OpTypeSampledImage typeof*{1}",
-                "%image = OpLoad typeof*{1} {1}",
-                "%sampler = OpLoad typeof*{2} {2}",
-                "%coord = OpLoad typeof*{3} {3}",
-                "%sampledImage = OpSampledImage %typeSampledImage %image %sampler",
-                "%result = OpImageSampleImplicitLod typeof*{0} %sampledImage %coord",
-                "OpStore {0} %result",
-                in(reg) &mut result,
-                in(reg) self,
-                in(reg) &sampler,
-                in(reg) &coord
-            );
-            result
-        }
-    }
-}
-
 #[allow(unused_attributes)]
 #[spirv(sampled_image)]
 #[derive(Copy, Clone)]
 pub struct SampledImage<I> {
     _image: I,
 }
 
-impl SampledImage<Image2d> {
-    pub fn sample(&self, coord: Vec2) -> Vec4 {
+impl<
+        const DIM: Dimensionality,
+        const DEPTH: ImageDepth,
+        const ARRAYED: bool,
+        const MULTISAMPLED: bool,
+        const SAMPLED: Sampled,
+        const FORMAT: ImageFormat,
+        const ACCESS_QUALIFIER: Option<AccessQualifier>,
+    >
+    SampledImage<
+        Image<
+            { DIM },
+            { DEPTH },
+            { ARRAYED },
+            { MULTISAMPLED },
+            { SAMPLED },
+            { FORMAT },
+            { ACCESS_QUALIFIER },
+        >,
+    >
+{
+    pub fn sample(&self, coord: impl ScalarOrVector) -> Vec4 {
```

Comment:
Shouldn't the coordinate type depend on the dimensionality of the image?

You can replace `trait ScalarOrVector` with a `trait ImageCoord<const DIM: Dimensionality>` you can implement for specific dimensionalities, though it might make even more sense to extract a single coordinate type, instead.

Though with a `ImageCoord` trait, we can support both `Vec3` and `Vec3A`, by converting the former to the latter.

Reply:
> Shouldn't the coordinate type depend on the dimensionality of the image?

Yes it should, just wasn't sure how to ensure that. That idea sounds great and I'll add it.